Web Extensions: Add delegate functions for bookmarks.getTree()

akutira-umich opened the pull request:
#### 50bba456612a2f91d3ee0804a13c800c23e9c6ae
<pre>
Web Extensions: Add delegate functions for bookmarks.getTree()
<a href="https://rdar.apple.com/155264420">rdar://155264420</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=295551">https://bugs.webkit.org/show_bug.cgi?id=295551</a>

Reviewed by Brian Weinstein.

Implementation of getTree and mockNode for create bookmarks for testing. Created a delegate for getTree and create.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::createParametersFromProtocolObject):
(WebKit::createParametersFromProtocolObjects):
(WebKit::WebExtensionContext::bookmarksCreate):
(WebKit::WebExtensionContext::bookmarksGetTree):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toAPI):
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getTree):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode initWithDictionary:]):
(-[_MockBookmarkNode identifierForWebExtensionContext:]):
(-[_MockBookmarkNode parentIdentifierForWebExtensionContext:]):
(-[_MockBookmarkNode titleForWebExtensionContext:]):
(-[_MockBookmarkNode urlStringForWebExtensionContext:]):
(-[_MockBookmarkNode indexForWebExtensionContext:]):
(-[_MockBookmarkNode childrenForWebExtensionContext:]):
(findParentInMockTree):
(-[TestBookmarksDelegate init]):
(-[TestBookmarksDelegate _webExtensionController:bookmarksForExtensionContext:completionHandler:]):
(-[TestBookmarksDelegate _webExtensionController:createBookmarkWithParentIdentifier:index:url:title:forExtensionContext:completionHandler:]):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureGetBookmarksDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetTree)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, CreateAndGetTree)):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:createBookmarkWithParentIdentifier:index:url:title:forExtensionContext:completionHandler:]):
(-[TestWebExtensionsDelegate _webExtensionController:bookmarksForExtensionContext:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/297971@main">https://commits.webkit.org/297971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f81238bd1608c5342d8b6ea6cdebbdb78286255f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113686 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33391 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23830 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64462 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115575 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33998 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86435 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41504 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116634 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27118 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102141 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66788 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20265 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63579 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96500 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20368 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123088 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40686 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30357 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95283 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41077 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98356 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95039 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24253 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40172 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17979 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40568 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46077 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40211 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43512 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41983 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

AriYoung00 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -111,6 +114,19 @@ WK_API_AVAILABLE(macos(15.4), ios(18.4), visionos(2.4))
+ */
+- (void)_webExtensionController:(WKWebExtensionController * _Nonnull)controller getBookmarksForExtensionContext:(WKWebExtensionContext * _Nonnull)context completionHandler:(void (^)(NSArray<NSObject<_WKWebExtensionBookmark> *> * _Nullable, NSError * _Nullable))completionHandler;
+
+- (void)_webExtensionController:(WKWebExtensionController * _Nonnull)controller createBookmark:(NSDictionary * _Nonnull)bookmarkData forExtensionContext:(WKWebExtensionContext * _Nonnull)context completionHandler:(void (^)(NSObject<_WKWebExtensionBookmark> * _Nullable, NSError * _Nullable))completionHandler;
```
This should also have a doc comment. Also, shouldn't it take a `_WKWebExtensionBookmark` for `bookmarkData`, rather than `NSDictionary`?

akutira-umich replied:
So the reason I think a NSDictionary is more appropriate here is because `_WKWebExtensionBookmark` is supposed to represent a bookmark tree node while `Create` should take in a createDetails object, which is a dictionary that only has parentId, title, url, and index.

AriYoung00 replied:
Then there should be another concrete type for a `createDetails` object which we pass up to the browser. Using an `NSDictionary` here means we don't get any help from the compiler in making sure that the fields which need to be present _are_ present, and of the correct type.

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,103 @@
+static WebExtensionBookmarksParameters createParametersFromProtocolObject(id<_WKWebExtensionBookmark> bookmark, WKWebExtensionContext *context)
+{
+    if (!bookmark)
+        return { };
```
It might be better to return a `std::optional<WebExtensionBookmarksParameters>` from here, rather than just returning an empty object on failure.

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,103 @@
+        completionHandler(Expected<WebExtensionBookmarksParameters, WebExtensionError> { WTFMove(parameters) });
+    });
+
+    [controllerDelegate _webExtensionController:controllerWrapper createBookmark:bookmarkData forExtensionContext:contextWrapper completionHandler:block.get()];
```
We should probably check whether the delegate responds to this selector (and other delegate selectors) before performing them. See `canProgrammaticallyOpenSidebar` in `WebExtensionContextAPISidebarCocoa.mm`.

AriYoung00 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -225,7 +333,79 @@
+
+    manager.get().internalDelegate.createBookmark = ^(NSDictionary *bookmarkData, void (^completionHandler)(NSObject<_WKWebExtensionBookmark> *, NSError *)) {
+        NSMutableDictionary *newBookmarkData = [bookmarkData mutableCopy];
+        NSString *newId = bookmarkData[@"id"] ?: [NSString stringWithFormat:@"%ld", (long)nextMockBookmarkId++];
```
Nit: I would split this into two lines:
```
        NSString *newId = bookmarkData[@"id"] ?: [NSString stringWithFormat:@"%ld", (long)nextMockBookmarkId];
        nextMockBookmarkId++;
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,103 @@
+
+    auto block = WTF::makeBlockPtr([contextWrapper, completionHandler = WTFMove(completionHandler)](NSArray<id<_WKWebExtensionBookmark>> *bookmarkNodes, NSError *error) mutable {
+        if (error || !bookmarkNodes)
+            return;
```
I might be misunderstanding something here, but wouldn't we want to pass the error back to the web process rather than just swallowing it?

akutira-umich replied:
Oh yes, that should be sent in a completionHandler.

AriYoung00 commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
```diff
@@ -0,0 +1,79 @@
+ @param context The context in which the web extension is running.
+ @return A string uniquely identifying this bookmark node.
+ */
+- (NSString *)identifierForWebExtensionContext:(WKWebExtensionContext *)context NS_SWIFT_NAME(identifier(for:));
```
I don't think this is necessarily a bad thing / something that should change, but just curious: Why do all of these methods take a `context` parameter?

b-weinstein replied:
I would be interested in the answer for this too.

akutira-umich replied:
Oh sorry I gave this answer in person to Ari. The idea behind it was that it follows the delegate pattern others have and that it makes the API flexible. More a quality of life thing than a neccesity

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,6 +67,32 @@
+    if (node.url) {
+        dictionary[urlKey] = node.url->createNSString().get();
+        if (node.url->createNSString().get().length)
+            dictionary[typeKey] = @"bookmark";
```
I think as it stands this will allow there to be nodes which are folders, but still contain the URL key corresponding to an empty string, which seems like something of a contradiction. I would suggest:
```suggestion
    if (node.url && !node.url->isEmpty()) {
        dictionary[urlKey] = node.url->createNSString().get();
        dictionary[typeKey] = @"bookmark";
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+    WKWebExtensionContext *contextWrapper = wrapper();
+
+    if (![controllerDelegate respondsToSelector:@selector(_webExtensionController:createBookmark:forExtensionContext:completionHandler:)]) {
+        completionHandler(makeUnexpected(@"Bookmarks delegate does not implement createBookmark."));
```
This error message should probably be a bit more generic:
```suggestion
        completionHandler(toWebExtensionError(apiName, nullString(), @"it is not implemented"));
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+
 void WebExtensionContext::bookmarksCreate(const std::optional<String>& parentId, const std::optional<uint64_t>& index, const std::optional<String>& url, const std::optional<String>& title, CompletionHandler<void(Expected<WebExtensionBookmarksParameters, WebExtensionError>&&)>&& completionHandler)
 {
 
```
```suggestion
    static NSString * const apiName = @"bookmarks.create()";
    
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+                completionHandler(toWebExtensionError(@"bookmarks.Create()", nullString(), @"bookmark was null or invalid"));
+                return;
+            }
+        }
```
Early return on error, invert success/error cases, use `apiName` constant:
```suggestion
        if (error) {
            completionHandler(toWebExtensionError(apiName, nullString(), @"error was reported"));
            return;
        }
        auto parametersOptional = createParametersFromProtocolObject(newBookmark, contextWrapper);

        if (!parametersOptional.has_value()) {
            completionHandler(toWebExtensionError(apiName, nullString(), @"bookmark was null or invalid"));
            return;
        }
        
        completionHandler(Expected<WebExtensionBookmarksParameters, WebExtensionError> { WTFMove(parametersOptional.value()) });
        }
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
 void WebExtensionContext::bookmarksGetTree(CompletionHandler<void(Expected<WebExtensionBookmarksParameters, WebExtensionError>&&)>&& completionHandler)
 {
-
+    ASSERT(isLoaded());
```
```suggestion
    static NSString * const apiName = @"bookmarks.getTree()";
    
    ASSERT(isLoaded());
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+    WKWebExtensionContext *contextWrapper = wrapper();
+
+    if (![controllerDelegate respondsToSelector:@selector(_webExtensionController:getBookmarksForExtensionContext:completionHandler:)]) {
+        completionHandler(makeUnexpected(@"Bookmarks delegate does not implement GetTree."));
```
```suggestion
        completionHandler(toWebExtensionError(apiName, nullString(), @"it is not implemented"));
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+
+    [controllerDelegate _webExtensionController:controllerWrapper getBookmarksForExtensionContext:contextWrapper completionHandler:^(NSArray<id<_WKWebExtensionBookmark>> *bookmarkNodes, NSError *error) {
+        if (error || !bookmarkNodes) {
+            completionHandler(toWebExtensionError(@"bookmarks.getTree()", nullString(), @"returned array of bookmarks was invalid"));
```
```suggestion
            completionHandler(toWebExtensionError(apiName, nullString(), @"returned array of bookmarks was invalid"));
```
It may also be worth considering breaking out `error` and `!bookmarkNodes` into their own cases, with different error messages.

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+        WebExtensionBookmarksParameters rootNode;
+
+        rootNode.nodeId = @"testBookmarksRoot";
+        rootNode.index = 0;
```
Do we still want to be using arbitrary values for these?

akutira-umich replied:
The nodeId value is for testing purposes. Index can be removed

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,6 +67,32 @@
+        dictionary[childrenKey] = childrenArray;
+    }
+
+    return [dictionary copy];
```
Is there a reason to copy the dictionary here?
```suggestion
    return dictionary;
```

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -208,7 +240,19 @@
+    WebProcess::singleton().sendWithAsyncReply(
+        Messages::WebExtensionContext::BookmarksCreate(parentmockId, index, url, title),
+        [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<WebExtensionBookmarksParameters, WebExtensionError>&& result) {
+            NSDictionary* newNodeDictionary = toAPI(result.value());
```
I believe this will panic if the result contains an error instead of a value. We should consider the error case and report it.

AriYoung00 commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -251,22 +295,18 @@
+        Messages::WebExtensionContext::BookmarksGetTree(),
+        [protectedThis = Ref { *this }, callback = WTFMove(callback)](Expected<WebExtensionBookmarksParameters, WebExtensionError>&& result) {
+            if (!result)
+                return;
```
We should be calling the callback with the error here, rather than just swallowing it

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+    if (parentId)
+        bookmarkData[@"parentId"] = parentId->createNSString().get();
+    if (index)
+        bookmarkData[@"index"] = @(*index);
```
Nit: prefer `.value()` over `operator*` for optionals
```suggestion
        bookmarkData[@"index"] = @(index.value());
```

AriYoung00 commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,118 @@
+    return createParametersFromProtocolObjects(@[bookmark], context)[0];
+}
+
+
```
If possible, I would invert these so that `createParametersFromProtocolObject` contains the implementation for a single object and `createParametersFromProtocolObjects` delegates out to the singular implementation for each node.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -111,6 +114,33 @@ WK_API_AVAILABLE(macos(15.4), ios(18.4), visionos(2.4))
+ taking the 4 parameters, creating the bookmark in its data store, and returning a new object that
+ represents the created item.
+ */
+- (void)_webExtensionController:(WKWebExtensionController * _Nonnull)controller createBookmarkWithParentIdentifier:(nullable NSString *)parentId index:(nullable NSNumber *)index url:(nullable NSString *)url title:(NSString * _Nonnull)title forExtensionContext:(WKWebExtensionContext * _Nonnull)context completionHandler:(void (^)(NSObject<_WKWebExtensionBookmark> * _Nullable, NSError * _Nullable))c ...
```
Is this a delegate function for `getTree`? Does `getTree` need to create bookmarks?

akutira-umich replied:
The reason create delegate is here is because in order to create mock bookmarks for the `TestWebKitAPI` it needs the delegate to be defined here

b-weinstein commented on Source/WebKit/Modules/OSX_Private.modulemap:
```diff
@@ -2074,6 +2074,11 @@ framework module WebKit_Private [system] {
     export *
   }
 
+  explicit module _WKWebExtensionBookmark {
```
Should this be `_WKWebExtensionBookmarks`?

akutira-umich replied:
The file is called `_WKWebExtensionBookmarks.h` but the protocol is called `_WKWebExtensionBookmark`, so I believe this is correct.

b-weinstein replied:
Is the module name based on the protocol though? I was under the impression the name of the module was based on the name of the header. @xeenon - do you have a thought here?

b-weinstein commented on Source/WebKit/Modules/iOS_Private.modulemap:
```diff
@@ -2980,6 +2980,11 @@ framework module WebKit_Private [system] {
     export *
   }
 
+  explicit module _WKWebExtensionBookmark {
```
Ditto this being `_WKWebExtensionBookmarks`.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -111,6 +114,33 @@ WK_API_AVAILABLE(macos(15.4), ios(18.4), visionos(2.4))
+ taking the 4 parameters, creating the bookmark in its data store, and returning a new object that
+ represents the created item.
+ */
+- (void)_webExtensionController:(WKWebExtensionController * _Nonnull)controller createBookmarkWithParentIdentifier:(nullable NSString *)parentId index:(nullable NSNumber *)index url:(nullable NSString *)url title:(NSString * _Nonnull)title forExtensionContext:(WKWebExtensionContext * _Nonnull)context completionHandler:(void (^)(NSObject<_WKWebExtensionBookmark> * _Nullable, NSError * _Nullable))c ...
```
Nit: I think we prefer `id <_WKWebExtensionBookmark>` for the `It takes the newly created bookmark node
 (as an object conforming to the `_WKWebExtensionBookmark` protocol)` argument.

b-weinstein replied:
Whoops, according to the style bot I was wrong here:

`Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:27: error: [spacing/objc-protocol] Protocol names shouldn't have a space before them.`

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
```diff
@@ -0,0 +1,79 @@
+typedef NS_ENUM(NSInteger, _WKWebExtensionBookmarkType) {
+    _WKWebExtensionBookmarkTypeBookmark,
+    _WKWebExtensionBookmarkTypeFolder,
+} NS_SWIFT_NAME(_WKWebExtension.BookmarkType) WK_API_AVAILABLE(macos(15.4), ios(18.4), visionos(2.4));
```
Nit: I think we should use the `TBA` macros instead of putting an actual version number in here.

1) We don't know when this will ship
2) I think this is already incorrect with the iOS 26 version number

b-weinstein replied:
Ditto this for all of the other parts of the delegate.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,126 @@
 }
 
-void WebExtensionContext::bookmarksCreate(const std::optional<String>& parentId, const std::optional<uint64_t>& index, const std::optional<String>& url, const std::optional<String>& title, CompletionHandler<void(Expected<WebExtensionBookmarksParameters, WebExtensionError>&&)>&& completionHandler)
+static Vector<WebExtensionBookmarksParameters> createParametersFromProtocolObjects(NSArray<id<_WKWebExtensionBookmark>> *, WKWebExtensionContext *);
```
NIt: I think we want a space after `id`.

`static Vector<WebExtensionBookmarksParameters> createParametersFromProtocolObjects(NSArray<id <_WKWebExtensionBookmark>> *, WKWebExtensionContext *);`

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,126 @@
-void WebExtensionContext::bookmarksCreate(const std::optional<String>& parentId, const std::optional<uint64_t>& index, const std::optional<String>& url, const std::optional<String>& title, CompletionHandler<void(Expected<WebExtensionBookmarksParameters, WebExtensionError>&&)>&& completionHandler)
+static Vector<WebExtensionBookmarksParameters> createParametersFromProtocolObjects(NSArray<id<_WKWebExtensionBookmark>> *, WKWebExtensionContext *);
+
+static std::optional<WebExtensionBookmarksParameters> createParametersFromProtocolObject(id<_WKWebExtensionBookmark> bookmark, WKWebExtensionContext *context)
```
Ditto a space after `id` here.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,126 @@
+
+static Vector<WebExtensionBookmarksParameters> createParametersFromProtocolObjects(NSArray<id<_WKWebExtensionBookmark>> *bookmarkNodes, WKWebExtensionContext *context)
 {
+    if (!bookmarkNodes)
```
This could be `if (!bookmarkNodes.count)` if we want to do the early return for an empty array too.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,126 @@
+}
+
+
+
```
A couple extra blank lines here.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -20,13 +24,126 @@
+        Vector<WebExtensionBookmarksParameters> topLevelNodes = createParametersFromProtocolObjects(bookmarkNodes, contextWrapper);
+        WebExtensionBookmarksParameters rootNode;
+
+        rootNode.nodeId = @"testBookmarksRoot";
```
I think there should be a `FIXME` comment about removing/changing this string value.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -63,6 +67,31 @@
+
+    if (node.parentId)
+        dictionary[parentIdKey] = node.parentId->createNSString().get();
+    dictionary[typeKey] = @"folder";
```
I think it would be more clear to put this in an `else` on line 84.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
```diff
@@ -140,6 +169,8 @@
     }
 }
 
+static int s_nextInternalBookmarkNodeId = 100;
```
What does this represent?

akutira-umich replied:
The purpose of this is that nodes aren't supposed to be something the user can assign, they are generated. I generated like this so that mock nodes in testing and in here could be consistent. So they each start at 100 and increment every time a new node is created.

b-weinstein commented on Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
```diff
@@ -70,6 +70,9 @@
 
 @property (nonatomic, copy) void (^didUpdateSidebar)(_WKWebExtensionSidebar *);
 
+@property (nonatomic, copy) void (^createBookmarkWithParentIdentifier)(NSString * parentId, NSNumber * index, NSString * url, NSString * title, void (^)(NSObject<_WKWebExtensionBookmark> *, NSError *));
```
Nit: Some extra spaces around the ` * `. It should be `NSString *parentId`, and the like for other ones.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -31,7 +31,111 @@
+    for (NSDictionary *childDict in childDictionaries)
+        [children addObject:[[_MockBookmarkNode alloc] initWithDictionary:childDict]];
+    return children;
+}
```
Nit: All of these methods should be on multiple lines.

```
- (NSString *)identifierForWebExtensionContext:(WKWebExtensionContext *)context
{ 
    return _dictionary[@"id"];
}
```

And should have the opening brace on the next line.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -31,7 +31,111 @@
+
+@interface _MockBookmarkNode : NSObject <_WKWebExtensionBookmark>
+- (instancetype)initWithDictionary:(NSDictionary *)dictionary;
+@property (nonatomic, strong) NSMutableDictionary *dictionary;
```
I'm not wild about this being backed by an `NSMutableDictionary`. Can it be backed by a regular NSDictionary? Or do callers get the dictionary from the node and modify it?

akutira-umich replied:
So I left it like this because it would be easier to add key value like `dateAdded` when I do `getRecent`

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -31,7 +31,111 @@
+
+
+@implementation TestBookmarksDelegate
+- (instancetype)init {
```
The `{` should be on a black line.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -31,7 +31,111 @@
+    if (url)
+        newBookmark[@"url"] = url;
+    NSString *newId = [NSString stringWithFormat:@"%ld", (long)self.nextMockBookmarkId];
+    _nextMockBookmarkId++;
```
Is there a reason to use `self.nextMockBookmarkId` and then `_nextMockBookmarkId++` on the next line?

Could it just be:
`NSString *newId = [NSString stringWithFormat:@"%ld", _nextMockBookmarkId++];`

akutira-umich replied:
I had it like this `NSString *newId = [NSString stringWithFormat:@"%ld", _nextMockBookmarkId++];` before. Ari told me to make it like this as a nit.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -189,27 +293,63 @@
+
+        newBookmarkData[@"index"] = index ?: @(uiProcessMockBookmarks.get().count);
+        [uiProcessMockBookmarks addObject:newBookmarkData];
+        completionHandler(adoptNS([[_MockBookmarkNode alloc] initWithDictionary:newBookmarkData]).get(), nil);
```
It looks like you use this same code in a number of tests. Can it be shared?

akutira-umich replied:
Yes it can, just changed this

b-weinstein reviewed (changes requested):
This is coming along! There are just enough changes that I'd like to see one more go.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
```diff
@@ -0,0 +1,81 @@
+    _WKWebExtensionBookmarkTypeFolder,
+} NS_SWIFT_NAME(_WKWebExtension.BookmarkType) WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_VISIONOS_TBA));
+
+/*! @abstract A class conforming to the ``_WKWebExtensionBookmark`` protocol represents a single bookmark node (a bookmark, folder, or separator) to web extensions. */
```
This comment says separator, but it doesn't look like we have any support for separators. Is this future work, or should the comment be removed?

akutira-umich replied:
Sorry at that point it was up in the air but now that comment should be removed

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
```diff
@@ -0,0 +1,81 @@
+ @param context The context in which the web extension is running.
+ @return The URL the bookmark points to. This should be `nil` for folders.
+ */
+- (nullable NSString *)urlForWebExtensionContext:(WKWebExtensionContext * _Nonnull)context NS_SWIFT_NAME(url(for:));
```
It's a bit confusing that the name is `urlForWebExtensionContext`, but it returns an `NSString`. Should this either:

- Return an NSURL?
- Be named something like `urlStringForWebExtensionContext`.

akutira-umich replied:
I think it makes more sense for it to be `urlStringForWebExtensionContext`. The MDN docs supports the idea that URL's are just strings

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h:
```diff
@@ -0,0 +1,81 @@
+ @param context The context in which the web extension is running.
+ @return An array of bookmark nodes contained within this folder. Should be `nil` if the node is not a folder.
+ */
+- (nullable NSArray<id <_WKWebExtensionBookmark>> *)childrenForWebExtensionContext:(WKWebExtensionContext * _Nonnull)context NS_SWIFT_NAME(children(for:));
```
I was wrong about the spacing here. It should be `id<`.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -31,7 +31,128 @@
+    if (self)
+        _dictionary = [dictionary mutableCopy];
+    return self;
+}
```
Nit: There should be blank lines after each closing brace.

akutira-umich committed:
Web Extensions: Add delegate functions for bookmarks.getTree()
rdar://155264420
https://bugs.webkit.org/show_bug.cgi?id=295551

Reviewed by Brian Weinstein.

Implementation of getTree and mockNode for create bookmarks for testing. Created a delegate for getTree and create.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionBookmarks.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::createParametersFromProtocolObject):
(WebKit::createParametersFromProtocolObjects):
(WebKit::WebExtensionContext::bookmarksCreate):
(WebKit::WebExtensionContext::bookmarksGetTree):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::toAPI):
(WebKit::toWebAPI):
(WebKit::WebExtensionAPIBookmarks::createBookmark):
(WebKit::WebExtensionAPIBookmarks::getTree):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(-[_MockBookmarkNode initWithDictionary:]):
(-[_MockBookmarkNode identifierForWebExtensionContext:]):
(-[_MockBookmarkNode parentIdentifierForWebExtensionContext:]):
(-[_MockBookmarkNode titleForWebExtensionContext:]):
(-[_MockBookmarkNode urlStringForWebExtensionContext:]):
(-[_MockBookmarkNode indexForWebExtensionContext:]):
(-[_MockBookmarkNode childrenForWebExtensionContext:]):
(findParentInMockTree):
(-[TestBookmarksDelegate init]):
(-[TestBookmarksDelegate _webExtensionController:bookmarksForExtensionContext:completionHandler:]):
(-[TestBookmarksDelegate _webExtensionController:createBookmarkWithParentIdentifier:index:url:title:forExtensionContext:completionHandler:]):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureGetBookmarksDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIMockNodeWithgetRecent)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPICreateParse)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIGetTree)):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, CreateAndGetTree)):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:createBookmarkWithParentIdentifier:index:url:title:forExtensionContext:completionHandler:]):
(-[TestWebExtensionsDelegate _webExtensionController:bookmarksForExtensionContext:completionHandler:]):

Canonical link: https://commits.webkit.org/297971@main
